src/stream/xsalsa20/xsalsa20_setup.c

```diff
@@ -0,0 +1,138 @@
+   @param st        [out] The destination of the XSalsa20 state
+   @param key       The secret key
+   @param keylen    The length of the secret key, must be 32 (octets)
+   @param nonce     The nonce
+   @param noncelen  The length of the nonce, must be 24 (octets)
+   @param rounds    Number of rounds (e.g. 20 for Salsa20)
+   @param subkey    [out] The subkey, NULL if not want a copy
+   @return CRYPT_OK if successful
+*/
+int xsalsa20_setup(salsa20_state *st,
+                   const unsigned char *key, unsigned long keylen,
+                   const unsigned char *nonce,    unsigned long noncelen,
+                   int rounds, unsigned char *subkey) {
+   unsigned char secondkey[32];
+   ulong32       x[64];                     /* input to & output fm doubleround */
+   int sti[] = {0, 5, 10, 15, 6, 7, 8, 9};  /* indices used to build subkey fm x */
+   int i = 0;
+
+   LTC_ARGCHK(st        != NULL);
+   LTC_ARGCHK(secondkey != NULL);
+   LTC_ARGCHK(key  != NULL);
+   LTC_ARGCHK(keylen    == 32);
+   LTC_ARGCHK(nonce     != NULL);
+   LTC_ARGCHK(noncelen  == 24);
+
+   if (rounds == 0) rounds = 20;
+   LTC_ARGCHK(rounds % 2 == 0); /* number of rounds must be evenly divisible by 2 */
+
+   LOAD32L(x[ 0], constants +  0);
+   LOAD32L(x[ 5], constants +  4);
+   LOAD32L(x[10], constants +  8);
+   LOAD32L(x[15], constants + 12);
+   LOAD32L(x[ 1], key +  0);
+   LOAD32L(x[ 2], key +  4);
+   LOAD32L(x[ 3], key +  8);
+   LOAD32L(x[ 4], key + 12);
+   LOAD32L(x[11], key + 16);
+   LOAD32L(x[12], key + 20);
+   LOAD32L(x[13], key + 24);
+   LOAD32L(x[14], key + 28);
+   LOAD32L(x[ 6], nonce +  0);
+   LOAD32L(x[ 7], nonce +  4);
+   LOAD32L(x[ 8], nonce +  8);
+   LOAD32L(x[ 9], nonce + 12);
+
+   /* use modified salsa20 doubleround (no final addition) */
+   _xsalsa20_doubleround(x, rounds);
+
+   for (i = 0; i < 8; ++i) {
+     STORE32L(x[sti[i]], secondkey + 4 * i);
+   }
+
+   /* this is equivalent to... */
+   LOAD32L(st->input[ 0], constants +  0);
+   LOAD32L(st->input[ 5], constants +  4);
+   LOAD32L(st->input[10], constants +  8);
+   LOAD32L(st->input[15], constants + 12);
+   LOAD32L(st->input[ 1], secondkey +  0);
+   LOAD32L(st->input[ 2], secondkey +  4);
+   LOAD32L(st->input[ 3], secondkey +  8);
+   LOAD32L(st->input[ 4], secondkey + 12);
+   LOAD32L(st->input[11], secondkey + 16);
+   LOAD32L(st->input[12], secondkey + 20);
+   LOAD32L(st->input[13], secondkey + 24);
+   LOAD32L(st->input[14], secondkey + 28);
+   LOAD32L(st->input[ 6], &(nonce[16]) + 0);
+   LOAD32L(st->input[ 7], &(nonce[16]) + 4);
+   st->input[ 8] = 0;
+   st->input[ 9] = 0;
+   /* ...this...
+       salsa20_setup(st, secondkey, 32, rounds);
+       salsa20_ivctr64(st, &nonce[16], 8, 0);
+   */
+   st->rounds = rounds;
+   st->ksleft = 0;
+   st->ivlen = 8;           /* set switch to say nonce/IV has been loaded */
+
+   /* copy out subkey if not NULL */
+   if (subkey != NULL) XMEMCPY(subkey, secondkey, sizeof(secondkey));
```

Comment:
Instead of copying why not work with the provided buffer if given?
```
unsigned char secondkey_buf[32], *secondkey = secondkey_buf;
...
if (subkey != NULL) secondkey = subkey;
...
#ifdef LTC_CLEAN_STACK
zeromem(secondkey_buf, ...
```

Reply:
implemented